src/app/CommandHandler.cpp

```diff
@@ -124,6 +118,10 @@ void CommandHandler::Close()
 {
     mSuppressResponse = false;
     MoveToState(CommandState::AwaitingDestruction);
+    // We must finish all async work before we can shut down a CommandHandler. The actual CommandHandler MUST finish their work in
+    // reasonable time or there is a bug.
+    VerifyOrDieWithMsg(mRefCount == 0, DataManagement, "CommandHandler::Close() called with %zu unfinished async work items",
+                       mRefCount);
```

Comment:
Continuing the [previous](https://github.com/project-chip/connectedhomeip/pull/11367#discussion_r742532680) conversation here (since Github seems to have closed out the previously outdated thread)

I think the key question here is whether the application that's holding a handle can rely on the underlying handler object always being valid, or if the rug can be pulled from underneath it because of a separate entity triggering a shutdown procedure.

I'd be fine with an approach that has the `CommandHandler` actually track the handle and 'null out' its innards on shutdown/destruction. Consequently, the application using the handle could still invoke the `->` operator, but it would results in a no-op because the object has ceased to exist.

Reply:
I start to believe that checking `mRefCount == 0` is problematic, because once `mRefCount == 0`, the object is already invalid, check a field of a invalid object make no sense at all. The underlying object can be already allocated to another instance.

The proper way we can ensure that the object is not being used after released is, never use the raw pointer, always use the reference counted handle.

Reply:
Maybe we should use a better name, since it is not a trivial refcount, it is a refcounter for async background work.